### PR TITLE
[WIP] Add CAN Method

### PR DESCRIPTION
**Pull request by YanisLalou**
Paper: https://arxiv.org/pdf/1901.00976
Mostly eq 3-4-5 + paragraph 3.4

### New Features:
* **CAN and CANLoss Implementation:**
  * Added `CANLoss` class to `skada/deep/_divergence.py` to implement the contrastive domain discrepancy (CDD) loss. 
  * Added `CAN` function to `skada/deep/_divergence.py` to implement the CAN domain adaptation method. 

### New Utilities:
* **SphericalKMeans:**
  * Added `SphericalKMeans` class to `skada/deep/utils.py` for clustering using cosine similarity.

### Testing:
* **New Tests for CAN:**
  * Added tests for CAN in `test_deep_divergence.py` to ensure the new method works as expected.

### Still needs to be done:
* Double check the implementation

### CODE REVIEW

File: skada/deep/losses.py [hunk 1]
```diff
@@ -189,6 +191,134 @@ def dan_loss(features_s, features_t, sigmas=None):
...
+    cluster_labels_t = cluster_labels_t[mask_t]
+
+    # Discard ambiguous classes
+    class_counts = torch.bincount(cluster_labels_t, minlength=n_classes)
+    valid_classes = class_counts >= class_threshold
+    mask_t = valid_classes[cluster_labels_t]
+    features_t = features_t[mask_t]
+    cluster_labels_t = cluster_labels_t[mask_t]
+
+    # Define sigmas
```
Comment (tgnassou) [inline skada/deep/losses.py:274]: Do you cannot use the mmd distance from DAN?

File: skada/deep/utils.py [hunk 2]
```diff
@@ -52,3 +54,170 @@ def check_generator(seed):
...
+
+        for _ in range(self.n_init):
+            if self.centroids is None:
+                centroids = self._init_centroids(X)
+            else:
+                centroids = self.centroids.to(self.device)
+
+            for n_iter in range(self.max_iter):
+                # Assign samples to closest centroids
+                similarities = self._compute_similarities(X, centroids)
```
[the remark this answers is not in the document]
Comment (YanisLalou) [reply skada/deep/utils.py]: agree

File: skada/deep/losses.py [hunk 3]
```diff
@@ -189,6 +191,134 @@ def dan_loss(features_s, features_t, sigmas=None):
...
+        features_t.unsqueeze(1), target_kmeans.cluster_centers_.unsqueeze(0)
+    )
+    mask_t = 0.5 * (1 - similarities.max(dim=1)[0]) < distance_threshold
+    features_t = features_t[mask_t]
+    cluster_labels_t = cluster_labels_t[mask_t]
+
+    # Discard ambiguous classes
+    class_counts = torch.bincount(cluster_labels_t, minlength=n_classes)
+    valid_classes = class_counts >= class_threshold
+    mask_t = valid_classes[cluster_labels_t]
```
[the remark this answers is not in the document]
Comment (YanisLalou) [reply skada/deep/losses.py:279]: 1. `class_counts = torch.bincount(cluster_labels_t, minlength=n_classes)`  counts how many samples are in each cluster.
2. `valid_classes = class_counts >= class_threshold` creates a boolean tensor where `True` indicates classes that have at least `class_threshold` samples.
3. `mask_t = valid_classes[cluster_labels_t]` is using the cluster labels as indices into the `valid_classes tensor. This create a boolean `mask_t`, where `True` indicates samples that belong to classes with enough representation.

This part of the code corresponds to the  `Filter the ambiguous classes` part of the paper pseudo algorithm.

File: skada/deep/losses.py (hunk 1 above)
Comment (YanisLalou) [reply skada/deep/losses.py:283]: The formula is not exactly the same as for the mmd since before computing each mean we apply a specific mask

File: skada/deep/losses.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tgnassou) [reply skada/deep/losses.py:279]: Got it!

File: skada/deep/losses.py [hunk 4]
```diff
@@ -189,6 +192,142 @@ def dan_loss(features_s, features_t, sigmas=None):
...
+        for c in range(n_classes):
+            mask = y_s == c
+            if mask.sum() > 0:
+                class_features = features_s[mask]
+                normalized_features = F.normalize(class_features, p=2, dim=1)
+                centroid = normalized_features.mean(dim=0)
+                source_centroids.append(centroid)
+
+    # Use source centroids to initialize target clustering
+    target_kmeans = SphericalKMeans(
```
Comment (tgnassou) [inline skada/deep/losses.py:257]: It should not be done also at the beginning of the epochs?

File: skada/deep/utils.py [hunk 5]
```diff
@@ -52,3 +54,170 @@ def check_generator(seed):
...
+                distances = 0.5 * (1 - similarities)
+                labels = torch.argmin(distances, dim=1)
+
+                # Update centroids
+                new_centroids = torch.zeros_like(centroids)
+                for k in range(self.n_clusters):
+                    if torch.any(labels == k):
+                        new_centroids[k] = X[labels == k].mean(dim=0)
```
[the remark this answers is not in the document]
Comment (tgnassou) [reply skada/deep/utils.py:170]: The `X` are still not normalized here?

File: skada/deep/utils.py [hunk 6]
```diff
@@ -52,3 +54,165 @@ def check_generator(seed):
...
+
+        for _ in range(self.n_init):
+            if self.centroids is None:
+                centroids = self._init_centroids(X)
+            else:
+                centroids = self.centroids.to(self.device)
+
+            for n_iter in range(self.max_iter):
+                # Assign samples to closest centroids
+                dissimilarities = self._compute_dissimilarities(X, centroids)
```
Comment (tgnassou) [inline skada/deep/utils.py:163]: There is a difference here with the function `cosine_similarities` de torch?

File: skada/deep/losses.py (hunk 4 above)
Comment (YanisLalou) [reply skada/deep/losses.py:257]: you're right

File: skada/deep/utils.py (hunk 6 above)
Comment (YanisLalou) [reply skada/deep/utils.py:163]: In paper: `cosine_dissimilarity` is  `0.5*(1 − cosine_similarity)`

File: skada/deep/utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (YanisLalou) [reply skada/deep/utils.py:173]: done

File: skada/deep/losses.py [hunk 7]
```diff
@@ -189,6 +191,134 @@ def dan_loss(features_s, features_t, sigmas=None):
...
+    n_classes = len(y_s.unique())
+
+    # Calculate source centroids
+    source_centroids = []
+    for c in range(n_classes):
+        mask = y_s == c
+        if mask.sum() > 0:
+            class_features = features_s[mask]
+            normalized_features = F.normalize(class_features, p=2, dim=1)
+            centroid = normalized_features.mean(dim=0)
```
[the remark this answers is not in the document]
Comment (YanisLalou) [reply skada/deep/losses.py]: In spherical k-means paper:
image

**Review by tgnassou — APPROVED** [review]
